src/DocExplorer/doctypes.ts

```diff
@@ -73,10 +75,40 @@ export type PatchworkDataType<D, T, V> = {
     docBefore: D,
     patches: A.Patch[]
   ) => Annotation<T, V>[];
+
+  /* Group annotations into logical units. This function get's passed all annotations
+   * that are not associated with any discussions
+   *
+   * The sort order is not preserved. For sorting implement the sortAnchorsBy method.
+   */
+  groupAnnotations?: (annotations: Annotation<T, V>[]) => Annotation<T, V>[][];
+
+  /* Resolves to the value the anchor is pointing to in a document.
+   * If the anchor cannot be resolved return undefined */
+  valueOfAnchor: (doc: D, anchor: T) => V | undefined;
+
+  /* Checks if two anchors overlap. This is used to associate edit annotations with
+   * discussions. A discussion grabs any annotations that overlap with the anchors
+   * associated with the discussion
+   *
+   * If this method is not implemented deep equal will be used as a fallback
+   */
+  doAnchorsOverlap?: (doc: D, anchor1: T, anchor2: T) => boolean;
+
+  /** Defines a value for each anchor that will be use to sort them by in descending order.
+   *  This is used for example in the SpatialSidebar to sort the annotation group.
+   *
+   *  If this method is not implemented the anchors will not be sorted.
+   */
+  sortAnchorsBy?: (doc: D, anchor: T) => any;
 };
```

Comment:
I should extract the anchor methods into an anchor class that each doc type implements. The thing that's still annoying is that we need to pass in the document to many of these, like doAnchorsOverlap because the cursor based anchor of MarkdownDoc can't be resolved otherwise